pkgs/build-support/build-fhs-userenv/chrootenv.c

```diff
@@ -20,56 +21,112 @@
 #include <sys/stat.h>
 #include <sys/wait.h>
 
-char *env_whitelist[] = {"TERM",
-                         "DISPLAY",
-                         "XAUTHORITY",
-                         "HOME",
-                         "XDG_RUNTIME_DIR",
-                         "LANG",
-                         "SSL_CERT_FILE",
-                         "DBUS_SESSION_BUS_ADDRESS"};
-
-char **env_build(char *names[], size_t len) {
-  char *env, **ret = malloc((len + 1) * sizeof(char *)), **ptr = ret;
-
-  for (size_t i = 0; i < len; i++) {
-    if ((env = getenv(names[i]))) {
-      if (asprintf(ptr++, "%s=%s", names[i], env) < 0)
-        errorf(EX_OSERR, "asprintf");
+#define LEN(x) sizeof(x) / sizeof(*x)
+
+char *env_blacklist[] = {};
+
+char **env_filter(char *envp[]) {
+  char **filtered_envp = malloc(sizeof(*envp));
+  size_t n = 0;
+
+  while (*envp != NULL) {
+    bool blacklisted = false;
+
+    for (size_t i = 0; i < LEN(env_blacklist); i++) {
+      if (!strncmp(*envp, env_blacklist[i], strlen(env_blacklist[i]))) {
+        blacklisted = true;
+        break;
+      }
+    }
+
+    if (!blacklisted) {
+      filtered_envp = realloc(filtered_envp, (n + 2) * sizeof(*envp));
+
+      if (filtered_envp == NULL)
+        errorf(EX_OSERR, "realloc");
+
+      filtered_envp[n++] = *envp;
     }
+
+    envp++;
+  }
+
+  filtered_envp[n] = NULL;
+  return filtered_envp;
+}
+
+void bind(char *from, char *to) {
+  if (mkdir(to, 0755) < 0)
+    errorf(EX_IOERR, "mkdir");
```

Comment:
it would be helpful here, which directory could have not been created in the error message:

```diff
-    errorf(EX_IOERR, "mkdir");
+    errorf(EX_IOERR, "could not create directory %s", to); 
```

Reply:
`errorf` macro handles that (errno, or `strerror` have that information).

Reply:
Which one do you find more informative?

```diff
--- main.c      2017-12-21 01:04:18.626520276 +0100
+++ main2.c     2017-12-21 01:06:39.371742074 +0100
@@ -1,15 +1,15 @@
 #define _GNU_SOURCE
 #include <errno.h>
 #include <error.h>
 #include <sysexits.h>
 #include <sys/stat.h>
 #define errorf(status, fmt, ...)                                               \
   error_at_line(status, errno, __FILE__, __LINE__, fmt, ##__VA_ARGS__)
 int main(int argc, char *argv[])
 {
   mkdir("/foo/bar", 0755);
-  errorf(EX_OSERR, "mkdir");
+  errorf(EX_OSERR, "failed to create directory '%s'", "/foo/bar");
   return 0;
 }
```

```diff
- ./main:main.c:13: mkdir: No such file or directory
+ ./main:main2.c:13: failed to create directory '/foo/bar': No such file or directory
```

Reply:
`mkdir` vs. `failed to create directory`: first one. It gives function that fails right out of the bat.

I do agree that having argument information would be very helpful, e.g. see `spitf`. In this case it would be:

```
errorf(EX_OSERR, "mkdir(%s)", "/foo/bar");
```

Reply:
So you expect people to know about low-level syscalls to understand error messages?

Reply:
While technically `mkdir` is a low-level syscall, everyone knows what `mkdir` does, or they would not be able to install NixOS in the first place :-)

Also, I don't think you'd be able to come up with a more helpful message for `unshare` than we currently have. In fact, if I were to remove `unshare` from error messages that currently have it, it would make debugging harder for me.

Also, using function name makes error much more googleable, which helps if it's user-recoverable.

`strerror` is supposed to generate user-friendly messages, and I think it's doing a pretty good job. `mkdir(/foo/bar): no such file or directory` is a user-friendly message.